book/visual-effects.md

```diff
@@ -803,39 +803,45 @@ class SaveLayer:
         canvas.restore()
 ```
 
-Now let's look at how we can add this to our existing `paint` method
-for `BlockLayout`s. Right now, this method draws a background and then
-recurses into its children, adding each drawing command straight to
-the global display list. Let's instead add those drawing commands to a
-temporary list first:
+Now let's look at how we can add this to our existing `paint` method for
+`BlockLayout`s. Now, _before_ we add its `cmds` command list to the overall
+display list, we can use `SaveLayer` to add transparency to the whole element.
+I'm going to do this in a new `paint_effects` method, which will wrap `cmds`
+in a `SaveLayer`. The actual `SaveLayer` will be computed in a new
+global `paint_visual_effects` method (because other object types will need it
+also).^[As part of adding this method, I've factored out a `self_rect` and
+`is_atomic` method from `paint`, you'll need to update that method also to call
+these helpers.]
 
 ``` {.python}
 class BlockLayout:
```

Comment:
Ok, I'm a bit confused. The reader will need to add `paint_effects` to every single layout class? Do all of them just call `paint_visual_effects`? I think we need to think this through carefully.

Reply:
They don't just blindly call it. For `BlockLayout` it conditions it on being atomic.

Reply:
I thought about this a lot on the drive home last night and I think it's better to correct for the atomic stuff inside `paint_tree` rather than in `BlockLayout`. For example, suppose you have an `<input>` element with `opacity`. We only want that `opacity` applied once, right? So the atomic inline thing actually applies, I think, to any `BlockLayout` whose root element is not a block element.

Maybe the right move here is to merge this and think about another pass later. But the reason I'm so worried about this PR is that I am worried that we have actual bugs here, and like usual I'm trying to achieve "obviously no bugs" and not yet succeeding.

Reply:
Ok I'll land it now and we can test & make another PR to try out moving is_atomic.

p.s. not sure what you mean by "whose root element is not a block element". Are you referring to it being for an inline element with some blocks inside it?